Skip `test_multithreaded_compute` on MoltenVK.

jimblandy opened the pull request:
Skip `tests::test_multithreaded_compute` in `examples/hello-compute/src/tests.rs` on MoltenVK for timeouts.

All commits should pass CI, but may be fine-grained enough that squashing makes sense. Up to you.

Various cleanups to expected failure code:

- Give test image snapshots distinctive names.

  Include the adapter name and driver name in the filenames generated to store the `-actual.png` and `-difference.png` files holding screenshots of test program output.

  Previously, the filename only incorporated the backend, but since we test the Vulkan backend on both Mac and Windows, one was overwriting the other in the artifacts.

- Upload GitHub workflow artifacts even when the test fails.

  When a test in `.github/workflows/ci.yml`'s `gpu-test` fails, ensure that it uploads the artifacts anyway. These are screenshots and comparison images, so they're valuable in debugging the failure.

- Allow expected test failures to match the driver name.

  Add a new field `FailureCase::driver`, which must match `AdapterInfo::driver` if given.

  Use this in `TestParameters::molten_vk_failure`.

- Implement `Default` for `FailureCase`, and use it where appropriate.

  This makes it easier to add new fields to `FailureCase` in the future, without having to fix every use.

- Make `TestParameters::specific_failure` take a `FailureCase`.

  Make `TestParameters::specific_failure` take a `FailureCase` struct as its sole argument, thus requiring callers to label each parameter with its meaning. This is clearer than using four positional parameters.

- Introduce TestParameters::backend_adapter_failure.

  Introduce a new `TestParameters` builder method, `backend_adapter_failure`, for marking an expected failure on a specific adapter and backend. Use where appropriate.

- Introduce TestParameters::adapter_failure_skip.

  Introduce a new `TestParameters` builder method for skipping a test on a specific adapter. Use where appropriate.

- Update doc comments for some `TestParameters` methods.

  Replace references to `specific_failures` with outdated parameter lists with plain English explaining what each function actually does.

- Use TestParameters::backend_failure method where appropriate.

**Checklist**

- [X] Run `cargo clippy`.
- [X] Run `cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
_Link to the issues addressed by this PR, or dependent PRs in other repositories_

**Description**
_Describe what problem this is solving, and how it's solved._

**Testing**
_Explain how this change is tested._

jimblandy committed:
Introduce TestParameters::adapter_failure_skip.

Introduce a new `TestParameters` builder method for skipping a test on
a specific adapter. Use where appropriate.

jimblandy committed:
Introduce TestParameters::backend_adapter_failure.

Introduce a new `TestParameters` builder method,
`backend_adapter_failure`, for marking an expected failure on a
specific adapter and backend. Use where appropriate.

jimblandy committed:
Update doc comments for some `TestParameters` methods.

Replace references to `specific_failures` with outdated parameter
lists with plain English explaining what each function actually does.

jimblandy committed:
Use TestParameters::backend_failure method where appropriate.

jimblandy committed:
Make `TestParameters::specific_failure` take a `FailureCase`.

Make `TestParameters::specific_failure` take a `FailureCase` struct as
its sole argument, thus requiring callers to label each parameter with
its meaning. This is clearer than using four positional parameters.

jimblandy committed:
Implement `Default` for `FailureCase`, and use it where appropriate.

This makes it easier to add new fields to `FailureCase` in the future,
without having to fix every use.

jimblandy committed:
Allow expected test failures to match the driver name.

Add a new field `FailureCase::driver`, which must match
`AdapterInfo::driver` if given.

Use this in `TestParameters::molten_vk_failure`.

jimblandy committed:
Skip `test_multithreaded_compute` on MoltenVK.

Skip `tests::test_multithreaded_compute` in `examples/hello-compute/src/tests.rs` on MoltenVK for timeouts.

jimblandy committed:
Give test image snapshots distinctive names.

Include the adapter name and driver name in the filenames generated to
store the `-actual.png` and `-difference.png` files holding
screenshots of test program output.

Previously, the filename only incorporated the backend, but since we
test the Vulkan backend on both Mac and Windows, one was overwriting
the other in the artifacts.

jimblandy committed:
Upload GitHub workflow artifacts even when the test fails.

When a test in `.github/workflows/ci.yml`'s `gpu-test` fails,
ensure that it uploads the artifacts anyway. These are screenshots and
comparison images, so they're valuable in debugging the failure.

cwfitzgerald commented on tests/src/lib.rs:
```diff
@@ -53,11 +53,80 @@ fn lowest_downlevel_properties() -> DownlevelCapabilities {
+    ///
+    /// If `false`, the test is run, and an unexpected pass is
+    /// reported if it doesn't fail.
+    pub skip: bool,
```
It would improve readability signifigantly to have this be an enum between "ExpectFailure" and "skip" or something like that, I've already gotten tripped up reading the bools in this MR.

`.molten_vk_failure(false)` sounds like it shouldn't fail on moltenvk, but molten_vk_failure(FailureMode::ExpectFailure) or whatever is much better

cwfitzgerald commented on tests/src/lib.rs:
```diff
@@ -53,11 +53,80 @@ fn lowest_downlevel_properties() -> DownlevelCapabilities {
+/// The default value of `FailureCase` applies to any test case (that
+/// is, there are no criteria to constrain the match), and the test is
+/// not skipped.
+pub struct FailureCase<S> {
```
We should be able to mark this as `#[non_exhaustive]` to really enforce the "you must use default"

cwfitzgerald reviewed (approved):
This is great stuff, have some nits, but good after

jimblandy committed:
Restrict FailureCase to holding static strings.

Use `&'static str` to represent adapter and driver names in
`FailureCase`. This requires us to convert to lower case each time we
match them against the adapter's strings, which is a heap allocation,
but this is not hot code, so it's better to be simple.

jimblandy committed:
Clarify marking tests that should be skipped / expeced to fail.

Give `TestParameters` two methods: `skip` and `expect_fail`,
both of which take a `FailureCase` as their sole argument,
indicating when the test should be skipped / expected to fail.

This means that the spine of each `TestParameters` builder expression
uses the word `skip` or `expect_fail`, rather than having a random
`bool` argument to some other method determine the difference.

Add a new method `FailureCase::applies_to(info)`, which decides
whether a given `FailureCase` applies to a given adapter running the
test.

Simplify logic for categorizing passes and failures, expected and
otherwise.

jimblandy commented:
Requesting a new review, since some of the new commits are non-trivial. None of the older commits are affected.

cwfitzgerald reviewed (approved):
Very nice!

cwfitzgerald committed:
Merge branch 'trunk' into test-multithreaded-compute-timeout-moltenvk
